### PR TITLE
New lint: vec_init_then_push

### PR DESCRIPTION
**Pull request by Jarcho**
fixes: #1483

This will trigger on `new`, `default`, and `with_capacity` when the given capacity is less than or equal to the number of push calls. Is there anything else this should trigger on?

changelog: Added lint: `vec_init_then_push`

### CODE REVIEW

File: clippy_lints/src/vec_init_then_push.rs [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    /// **Why is this bad?** The `vec![]` macro is both more performant and easier to read than
+    /// multiple `push` calls.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// let mut v: Vec<u32> = Vec::new();
+    /// v.push(0);
```
Comment (llogiq) [inline clippy_lints/src/vec_init_then_push.rs:24]:
```suggestion
    /// let mut v = Vec::new();
    /// v.push(0_u32);
```

File: clippy_lints/src/vec_init_then_push.rs [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let v: Vec<u32> = vec![0];
```
Comment (llogiq) [inline clippy_lints/src/vec_init_then_push.rs:28]:
```suggestion
    /// let v = vec![0_u32];
```

**Review by llogiq — APPROVED** [review]

**Comment by llogiq** [comment]
I like this addition! Just a small nit: I'd make use off inference to keep the docs lean.

Also I think the Changelog line in the PR description should be the last line.

File: clippy_lints/src/vec_init_then_push.rs (hunk 1 above)
Comment (Jarcho) [reply clippy_lints/src/vec_init_then_push.rs]: Fixed.

File: clippy_lints/src/vec_init_then_push.rs (hunk 2 above)
Comment (Jarcho) [reply clippy_lints/src/vec_init_then_push.rs]: Fixed.

**Comment by llogiq** [comment]
Thank you! @bors r+

**Pull request merged**